### PR TITLE
Hevc_enc: based on media studio VA

### PR DESCRIPTION
**Pull request by wudping**
enable hevc encoding based on media studio VA.

### CODE REVIEW

**Comment by wudping** [comment]
please help to review. Thanks in advance.

File: interface/VideoCommonDefs.h [hunk 1]
```diff
@@ -42,6 +42,10 @@
 #define StaticPointerCast std::static_pointer_cast
 #define EnableSharedFromThis std::enable_shared_from_this
 
+#ifdef __ENABLE_MEDIA_STUDIO_VA__
```
Comment (xuguangxin) [inline interface/VideoCommonDefs.h]: please move this to configure.ac.

File: encoder/vaapiencoder_base.cpp [hunk 2]
```diff
@@ -574,10 +582,26 @@ bool VaapiEncoderBase::initVA()
...
+        It's necessary to pass two attribute:
+        VAConfigAttribRTFormat and VAConfigAttribRateControl to libva;
+        The value of VAConfigAttribRateControl should be 0x82 not RATE_CONTROL_CBR;
+        Or else, HEVC encoding will end up with an error: attribute not supported.
+        */
+        if (RATE_CONTROL_CBR == m_videoParamCommon.rcMode) {
+            attrib[0].type = VAConfigAttribRTFormat;
+            attrib[0].value = 0;
+            attrib[1].type = VAConfigAttribRateControl;
+            attrib[1].value = 0x82; //RATE_CONTROL_CBR
```
Comment (xuguangxin) [inline encoder/vaapiencoder_base.cpp]: could you find where we need this magic number, is it another RATE_CONTROL_XXX?

**Review by xuguangxin — CHANGES_REQUESTED** [review]

File: encoder/vaapiencoder_base.cpp (hunk 2 above)
Comment (wudping) [reply encoder/vaapiencoder_base.cpp]: I can't find out any MACRO for it.

Comment (xuguangxin) [reply encoder/vaapiencoder_base.cpp]: is it a VA_RC_MB|VA_RC_CBR ?

Comment (wudping) [reply encoder/vaapiencoder_base.cpp]: I get your mean, thanks for your help.

**Comment by wudping** [comment]
updated, please help to review, thanks.

**Review by xuguangxin — APPROVED** [review]

File: encoder/vaapiencoder_base.cpp [hunk 3]
```diff
@@ -574,10 +582,26 @@ bool VaapiEncoderBase::initVA()
...
-        pAttrib = &attrib;
+        attrib[0].type = VAConfigAttribRateControl;
+        attrib[0].value = m_videoParamCommon.rcMode;
+        pAttrib = attrib;
         attribCount = 1;
+#ifdef __ENABLE_HEVC_ENC_ON_STUDIO_VA__
+        /*
+        It's necessary to pass two attribute:
+        VAConfigAttribRTFormat and VAConfigAttribRateControl to libva;
+        The value of VAConfigAttribRateControl should be 0x82 not RATE_CONTROL_CBR;
```
Comment (lizhong1008) [inline encoder/vaapiencoder_base.cpp]: the comments of 0x82 should should be changed to be  VA_RC_MB | VA_RC_CBR

File: configure.ac [hunk 4]
```diff
@@ -395,6 +395,16 @@ AC_ARG_ENABLE([tests],
 
 AM_CONDITIONAL(ENABLE_TESTS, test "$enable_tests" = "yes")
 
+AC_ARG_ENABLE(media-studio-va,
+    [AC_HELP_STRING([--enable-media-studio-va],
+        [enable being based on media studio libva @<:@default=no@:>@])],
+    [], [enable_media_studio_va="no"])
+
+if test "$enable_media_studio_va" = "yes"; then
+    AC_DEFINE([__ENABLE_HEVC_ENC_ON_STUDIO_VA__], [1],
```
Comment (lizhong1008) [inline configure.ac]: would better change it to be __ENABLE_STUDIO_VA__, there is an existed option of  BUILD_H265_ENCODER_. And if hevc mentioned in the file, please use a unified name  "H265", not HEVC.

**Comment by wudping** [comment]
updated, please help to review.

**Comment by wudping** [comment]
updated, please help to review, thanks.

**Pull request merged**